### PR TITLE
DM-43101: Add exposure.can_see_sky metadata field

### PR DESCRIPTION
**Pull request by timj**
No attempt is made to generate a value when importing from an old universe. (replaces #972)

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [x] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

### CODE REVIEW

File: python/lsst/daf/butler/timespan_database_representation.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ class _SpecialTimespanBound(enum.Enum):
     """
 
 
-TimespanBound = Union[astropy.time.Time, _SpecialTimespanBound, None]
+TimespanBound: TypeAlias = astropy.time.Time | _SpecialTimespanBound | None
```
Comment (timj) [inline python/lsst/daf/butler/timespan_database_representation.py]: I'm not sure why this code is copied from `_timespan.py`.

File: python/lsst/daf/butler/logging.py [hunk 2]
```diff
@@ -292,18 +292,18 @@ class ButlerLogRecords(_ButlerLogRecords):
...
+    def from_records(cls, records: Iterable[ButlerLogRecord]) -> Self:
         """Create collection from iterable.
 
         Parameters
         ----------
         records : iterable of `ButlerLogRecord`
             The records to seed this class with.
         """
-        return cls.model_construct(root=list(records))
+        return cast(Self, cls.model_construct(root=list(records)))
```
Comment (timj) [inline python/lsst/daf/butler/logging.py]: `cls.model_construct()` returns `ButlerLogRecords` and mypy can't work out that's not `Self`.

File: python/lsst/daf/butler/timespan_database_representation.py (hunk 1 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/timespan_database_representation.py]: It was just a mistake made when refactoring one module into two modules.  We should remove it and import instead it if it's actually used here.

File: python/lsst/daf/butler/json.py [hunk 3]
```diff
@@ -42,7 +44,7 @@ class SupportsSimple(Protocol):
     serialization using "simple" methods names.
     """
 
-    _serializedType: Type
+    _serializedType: type[BaseModel]
```
Comment (TallJimbo) [inline python/lsst/daf/butler/json.py]: Should actually by `ClassVar[type[BaseModel]]`, I suspect.

Aside: this is correct for the code as it is now, but I think we really want to support things that are recognized by Pydantic other than `BaseModel` (like `typing.TypedDict`) as "simple" objects in the future.  I think that would involve making this into:

```
class SupportsSimple(Protocol, Generic[_T]):

    _adapter: ClassVar[pydantic.TypeAdapter[_T]]
```

**Review by TallJimbo — APPROVED** [review]
I'd prefer to revert the `Union -> |` commit and tell Ruff to either ignore that rule (ideal) or ignore at least the majority of those lines, as I think `Union` is better for anything multi-line (or potentially multi-line in the future).

I'm also not sold on the `Self` commit, as that is a subtle definition change rather than just notation, when inheritance is in play; I prefer not to use `Self` unless I want to say that derived classes must return the most-derived type.  Looks like we could be using `from __future__ import annotations` in `ButlerLogRecords` instead, though - probably dates from when pydantic couldn't handle that.

File: python/lsst/daf/butler/json.py (hunk 3 above)
Comment (timj) [reply python/lsst/daf/butler/json.py]: Maybe, but `TypeAdapter` doesn't support `model_dump_json` which is what the code relies on (it's called `dump_json` with `TypeAdapter`).

I'll do your `ClassVar` fix.

**Pull request merged**